### ISSUE TITLE
Improve experience of checking in node_modules

### ISSUE THREAD

**Issue by damccorm** [actions/javascript-template#4]
Right now, we use husky but its kinda painful. In particular, for users who use editors like VSCode, not having node_modules in the gitignore causes some problems.

We've thought about the idea of having master (or another branch) basically be a dev branch, with no node_modules checked in and then creating an action to automatically create release branches. Definitely open to suggestions here as well.

**Comment by Embraser01** [comment]
A solution could be to use the [Yarn Zero-Install](https://yarnpkg.github.io/berry/features/zero-installs) feature in Yarn v2! Even if it's not released yet, it's quite stable.

With it, dependencies are stored in a single folder as zipfiles, no need to keep a `node_modules` folder 😃 and any commit is ready to be run. 

Drawbacks:
- Editor compatibility (VSCode has a plugin, but Webstorm need node_modules 🙁)
- Forced to use Yarn over NPM

I could make a PR if you want to give it a try!

**Comment by damccorm** [comment]
I think what we've settled on (and this is reflected in the current state of this repo/linked docs) is not adding node_modules to master and only installing node_modules on release branches. This also encourages good patterns of having a release branch to pin to.

I think the editor concerns you listed are significant enough, along with it still being in preview, that I probably want to stick with that solution (and I think it works pretty well so I don't see a huge need to switch). With that said, the zero-install feature looks really cool and I'm definitely intrigued - I definitely want to try it out.

Also note, I want to leave this issue open until we update our first party actions to use the pattern I just laid out, so I think it should stay open for now.

**Comment by ajafff** [comment]
Some thoughts:
* Does this work for native modules? You will get a different binary depending on your local environment (OS, Node.js version, processor, ...). These aren't guaranteed to work in the target environment. Does this mean native modules can only be used in Docker actions?
* Without an `npm install` you will never get new patch releases of dependencies unless you publish a new version of your action
* checking in `node_modules` is only part of the story: when using a transpiled language (such as TypeScript) you need to push the transpiled files as well. Typically these are listed in `.gitignore`.

**Comment by damccorm** [comment]
> Does this work for native modules?

Not super strong in this area, but given what you've described I guess not. With that said, it seems like the right choice - I don't really think we want to be building those at runtime. In general, that's the pattern we're trying to avoid - having every action in your workflow build at runtime and adding a bunch of time to every job.

> Without an npm install you will never get new patch releases of dependencies unless you publish a new version of your action

Yep, this is by design. A released version should remain static for security reasons (e.g. if someone were to maliciously update a dependency, we don't slide without an explicit choice)

> checking in node_modules is only part of the story: when using a transpiled language (such as TypeScript) you need to push the transpiled files as well. Typically these are listed in .gitignore.

Yep. We don't have those in our .gitignore, and that's part of our guidance docs.

**Comment by moltar** [comment]
> In general, that's the pattern we're trying to avoid

I don't think you can avoid that, especially if you want for actions to work across different OSes.

Native modules won't work even on the same OS across different versions sometimes, or different versions of dependencies (e.g. header files) not to mention an entirely different OS.

One modules comes to mind that often needs rebuilding: `node-sass`. It seems like every time I do `brew update` or macOS updates all of my repos need local rebuilds.

Maybe packaging actions as Docker images would work better. But then they won't be able to run natively on an OS, but I am not sure if this is a requirement.

**Comment by damccorm** [comment]
I think in general, my guidance would be to use a docker action if you _need_ native modules. If that absolutely doesn't work for you, you could install your own dependencies at runtime (e.g. exec an npm install from node, or exec a script that does that) - that's not a pattern we recommend though simply because of the time that adds on to execution.

**Comment by warrenbuckley** [comment]
@damccorm have you sen this repo/action to help automate this - https://github.com/bumfo/action-deploy

Perhaps something to consider for template use?

**Comment by radu-matei** [comment]
How about using the package registry to distribute them?
This way, you don't have to push `node_module`, and could potentially handle OS-specific dependencies (same way they work right now, if I pull from OS/Arch, I get the correct deps).

**Comment by chunlea** [comment]
Why not we use some tools like `gulp` or `webpack` to build the `dist`, aka `lib`, and copy the code we need from `node_module`, rather than checkin the production `node_module`?

**Comment by damccorm** [comment]
> How about using the package registry to distribute them?
> Why not we use some tools like gulp or webpack
We're actually reconsidering this and thinking about both of these/similar solutions. Note that using the GitHub Package Registry isn't supported (yet), but you can use something like `webpack` currently.

**Comment by alex-page** [comment]
Thanks @swinton my projects `package.json` scripts looked like:
```json
    "watch": "ncc build index.js --watch --minify",
    "build": "ncc build index.js --minify",
```

**Comment by thomaseizinger** [comment]
Have you considered requiring actions to attach specific files to a release?
The CI configuration of an action could build an executable file that is attached to the release and downloaded by the runner.
If that is not available, it would fallback to downloading the repositories snapshot.

**Comment by bryanmacfarlane** [comment]
Update:

1. We moved away from husky and updated the docs on how to create a coding branch without node_modules and a distribution branch with prod modules.  The bare necessities to run an action from the github graph.
2. We have future work planned to integrate with GPR as a distribution mechanism (so repo has no branch with prod modules).   We evaluated releases (and tar gzs attached) but opt'ed for the better immutability benefits of a gh sha or a GPR type.

Just know that the feedback is heard loud and clear and we're investing in this area.  

I'm going to close since we're tracking the GPR work.

Thanks so much for the feedback and the time to write this up.
